### ISSUE TITLE
Weirdness with default params in Lua lib methods

### ISSUE THREAD

**Issue by ThunderAxe31** [TASVideos/BizHawk#2098]
According to the BizHawk Lua functions [documentation](http://tasvideos.org/Bizhawk/LuaFunctions.html), the function should work as the following:
```
void gui.drawPolygon(nluatable points, [int? offsetx = null], [int? offsety = null], [color? line = null], [color? background = null])
```
However, I've encountered two issues. First, there seem to not exist offsetx and offsety arguments.
Second, the function doesn't work at all if you use integers as color values.
For example, the following doesn't work (nothing appears to be drawn of screen):
```
gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, 0xFF00FF00, 0xFF0000FF)
```
However, the following **does** work:
```
gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, "Green", "Blue")
```
And if you're wondering if the integers I used are valid, yes, they are, because they work for other functions like gui.drawLine.

**Comment by YoshiRulz** [comment]
Those parameters definitely exist (2.4.2, 2.5 dev at 81628bf10):
https://github.com/TASVideos/BizHawk/blob/546ccda9189e346840909350eca8b1fdcadad081/BizHawk.Client.EmuHawk/tools/Lua/Libraries/GuiLuaLibrary.cs#L125
https://github.com/TASVideos/BizHawk/blob/81628bf1098ab92681e7d9ea6a2f216656d927c0/src/BizHawk.Client.EmuHawk/tools/Lua/Libraries/GuiLuaLibrary.cs#L125

I'm not able to test right now, but this will probably work: `gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, nil, nil, 0xFF00FF00, 0xFF0000FF)`.

**Comment by ThunderAxe31** [comment]
Nope. I get the following:
`callNLua.Exceptions.LuaScriptException: [string "main"]:4: invalid arguments to method call`

However, this worked:
`gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, 0, 0, 0xFF00FF00, 0xFF0000FF)`

It's just that it refuses any kind of null value. Also, for some weird reason, string values are accepted and used as color names, as in my example in the first comment.

**Comment by YoshiRulz** [comment]
Okay, so your first call doesn't appear to work because the parameters are passed like this:
```lua
gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, 0xFF00FF00, 0xFF0000FF)
-- points = {{10,10},{50,10},{10,50}}
-- offsetX = -16711936
-- offsetY = -16776961
-- line = null
-- background = null
```
It is actually working, but the poly is translated to be way outside the canvas boundary. Your second call works because ??? who knows, but the parameters somehow end up like this:
```lua
gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, "Green", "Blue")
-- points = {{10,10},{50,10},{10,50}}
-- offsetX = null
-- offsetY = null
-- line = new Color("Green")
-- background = new Color("Blue")
```
Similarly, your third call:
```lua
gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, 0, 0, 0xFF00FF00, 0xFF0000FF)
-- points = {{10,10},{50,10},{10,50}}
-- offsetX = 0
-- offsetY = 0
-- line = new Color(-16711936)
-- background = new Color(-16776961)
```
This `Color` construction is actually in the docs you linked.
https://github.com/TASVideos/BizHawk/blob/ad16294b11a4e12c9f41d88aa100210c498bea26/src/BizHawk.Client.Common/lua/LuaDocumentation.cs#L33-L34

In summary, everything is working as normal **except** optional (i.e. nullable) `int` parameters. Your second call by all accounts should not work.

**Comment by ThunderAxe31** [comment]
Ok, then I see the function works fine. Actually, it works too much well, since my second example call shouldn't be supposed to work, in theory. I guess we may close this issue now, unless you deem this quirk worthy of being analyzed.

**Comment by YoshiRulz** [comment]
@adelikat There's a genuine bug in here, default params aren't bridged to Lua correctly.

**Comment by ShinobiWannabe** [comment]
Hello, would like to add that I too have encountered this issue with:

**gui.drawText( 16, 32, "Some message", 0x7F0000FF, 0x00007FFF, 8, "Arial Narrow", "bold", "center", "middle" )**

based on the documentation it should be working.

//**gui.drawText(int x, int y, string message, [color? forecolor = null], [color? backcolor = null], [int? fontsize = null], [string fontfamily = null], [string fontstyle = null], [string horizalign = null], [string vertalign = null])**

The message without the optional params works fine.

**gui.drawText( 0, 0, "Some message" );**

**Comment by ShinobiWannabe** [comment]
Ah ok, I stepped through it in VS saw why i was getting the error.  That font is not installed my system.  I exchanged "Arial Narrow" for "Calibri", and it worked. The backcolor is probably not set either because the text is all by itself on the emulation window maybe. Location was in  new FontFamily(fontfamily) with the function DrawString in GuiApi.cs 

Anyway this should probably be closed for now.  Probably will find something else, but next time I will step through it.

**Comment by YoshiRulz** [comment]
There are still actual bugs here. I'll check these on `master` when I'm next able to, but I don't think anything has changed since 2.4.2.
* doesn't work but should: `gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, nil, nil, 0xFF00FF00, 0xFF0000FF)`
* works but shouldn't: `gui.drawPolygon({ {10, 10}, {50, 10}, {10, 50} }, "Green", "Blue")`
